### DISCUSSION TITLE
decrypt a file encrypted with Openssl

### DISCUSSION THREAD

**Discussion by priore** [krzyzanowskim/CryptoSwift#960] [General]
- CryptoSwift version: last
- Swift version: 4.1
- Platform (iOS/macOS/Linux/...): iOS 
- Installation method (CocoaPods/Carthage/SPM/...): cocoapods
- Project name: Test

# Description

When I try to decrypt a file, it is a further encrypted certificate with Openssl, look at the last openssl command I've executed, it returns the wrong content.

Many thanks for your support.

# Steps to reproduce

**create certificates with openssl :**

_openssl genrsa -aes256 -out private.pem 2048 -days 3000_
_openssl rsa -pubout -outform PEM -in private.pem -out public.pem_

**crypt public certificate with openssl :**

_openssl enc -aes-256-ecb -md md5 -nosalt -base64 -in public.pem -out public.enc -pass pass:12345678901234567890123456789012_

Sample code to reproduce the issue:

```swift

        do {
            
            guard let path = Bundle.main.path(forResource: "public", ofType: "enc") else {
                return nil
            }
            
            let public_enc = try String(contentsOfFile: path, encoding: .utf8)
            let data_enc = Data(base64Encoded: public_enc, options: .ignoreUnknownCharacters)
            let key = "12345678901234567890123456789012".bytes.md5()
            let aes = try AES(key: key, blockMode: .ECB, padding: .pkcs7)
            let pubkey = try data_enc!.decrypt(cipher: aes)
            let s = String(data: pubkey, encoding: String.Encoding.ascii)

            return s
            
        } catch {
            debugPrint(error)
            return nil
        }
```

**Comment by krzyzanowskim** [comment]
Before I start looking at this, did you read the historic Q https://github.com/krzyzanowskim/CryptoSwift/issues?utf8=%E2%9C%93&q=is%3Aissue+sort%3Aupdated-desc+openssl ? I'm giving the recipe how to build OpenSSL command-line compatible parameters somewhere.

**Comment by priore** [comment]
yes before writing I read all the issues containing the word Openssl, I have tried some solutions but none to be valid for a file encrypted with openssl and saved in base64.

a feature of the openssl files is that they contain characters that are not accepted for base64encoded, need to use the additional option ignoreUnknownCharacters, like below:

```swift
let data_enc = Data(base64Encoded: public_enc, options: .ignoreUnknownCharacters)
````

and I do not understand if this is the problem.

**Comment by priore** [comment]
nothing, i have try without base64 encoded file, only binary file, but have same incorrect result :

```
openssl enc -aes-256-ecb -md md5 -nosalt -in public.pem -out public.enc -pass pass:12345678901234567890123456789012
```

```swift
            let url = Bundle.main.url(forResource: "public", withExtension: "enc")
            let public_enc = try Data(contentsOf: url!)
            let key = "12345678901234567890123456789012".bytes.md5()
            let aes = try AES(key: key, blockMode: .ECB)
            let pubkey = try public_enc.decrypt(cipher: aes)
            let s = String(data: pubkey, encoding: String.Encoding.ascii)
            print(s)
```

i appreciate your help.

**Comment by krzyzanowskim** [comment]
will work if u use key directly with openssl (not password):

```
openssl enc -aes-256-ecb -md md5 -nosalt -in public.pem -out public.enc -K 3132333435363738393031323334353637383930313233343536373839303132
```

`3132333435363738393031323334353637383930313233343536373839303132` is `"12345678901234567890123456789012"` in hex, -K is a key value, the same that is used for AES instance.

```swift
let aes = try AES(key: [UInt8](hex: "3132333435363738393031323334353637383930313233343536373839303132"), blockMode: ECB(), padding: .noPadding)
let pubkey = try public_enc.decrypt(cipher: aes)

```

According to OpenSSL docs, you have to figure out how the key is derived from the password.

https://www.openssl.org/docs/man1.0.2/crypto/EVP_BytesToKey.html

> If the total key and IV length is less than the digest length and MD5 is used then the derivation algorithm is compatible with PKCS#5 v1.5 otherwise a non standard extension is used to derive the extra data.
